flink-runtime/src/main/java/org/apache/flink/runtime/security/token/KerberosDelegationTokenManager.java

```diff
@@ -110,13 +139,62 @@ public void obtainDelegationTokens(Credentials credentials) {
      * task managers.
      */
     @Override
-    public void start() {
-        LOG.info("Starting renewal task");
+    public void start() throws Exception {
+        checkNotNull(scheduledExecutor, "Scheduled executor must not be null");
+        checkNotNull(executorService, "Executor service must not be null");
+        checkState(tgtRenewalFuture == null, "Manager is already started");
+
+        if (!kerberosRenewalPossibleProvider.isRenewalPossible()) {
+            LOG.info("Renewal is NOT possible, skipping to start renewal task");
+            return;
+        }
+
+        startTGTRenewal();
+    }
+
+    private void startTGTRenewal() throws IOException {
+        LOG.debug("Starting credential renewal task");
+
+        UserGroupInformation currentUser = UserGroupInformation.getCurrentUser();
+        if (currentUser.isFromKeytab()) {
+            // In Hadoop 2.x, renewal of the keytab-based login seems to be automatic, but in Hadoop
+            // 3.x, it is configurable (see hadoop.kerberos.keytab.login.autorenewal.enabled, added
+            // in HADOOP-9567). This task will make sure that the user stays logged in regardless of
+            // that configuration's value. Note that checkTGTAndReloginFromKeytab() is a no-op if
+            // the TGT does not need to be renewed yet.
+            long tgtRenewalPeriod = configuration.get(KERBEROS_RELOGIN_PERIOD).toMillis();
+            tgtRenewalFuture =
+                    scheduledExecutor.scheduleAtFixedRate(
+                            () ->
+                                    executorService.execute(
+                                            () -> {
+                                                try {
+                                                    LOG.debug("Renewing TGT");
+                                                    currentUser.checkTGTAndReloginFromKeytab();
```

Comment:
Can you please write a test case for the periodic renewal? (`ManuallyTriggeredScheduledExecutorService` is usually pretty helpful for this)

The main reason I'm bringing this up that this still uses Hadoop's UGI internals, which makes this mostly untestable. My feeling is that `KerberosRenewalPossibleProvider` should become something that completely hides the Hadoop internals from DTM (eg. `KerberosClient` with two implementations: `HadoopKerberosClient` and `TestingKerberosClient` -> see eg. `TestingDeclarativeSlotPool` for how the "testing" implementations in Flink are usually created).

Reply:
I've started to have a look...

Reply:
@dmvk did you have a chance to consider the situation in-depth? I've my own suggestion which may or may not intersect your opinion. Namely if it's not too horror complex upgrading to mockito 3.4.0 and choosing bullet point 1 is my preference. Though no idea why mockito is super old, maybe there was no agreement to upgrade that?! If you have some insights please share.

If that would be an overkill then I would vote on bullet point 3 because all the other options would add hard to maintain and brittle solutions.

WDYT?

Reply:
In the meantime I've had a deeper look at the mock framework versions and [here](https://gist.github.com/gaborgsomogyi/4e8b120cbebde6d2a6903e5fcccbbaff) is the extract.

I've added a mockito version upgrade based unit test which we may or may not agree on.
This single test works but not sure about all other tests, waiting on jenkins to show it to us.
If we say this is not the direction or all other tests are not working then I can roll it back easily.
As a general note from my side. Even if we would choose this solution I think it would be good to split it up to 3 PRs:
* This PR w/o startTGTRenewal test
* Upgrade mockito version
* Add startTGTRenewal test

If this not works or you disagree plz suggest a way.

Reply:
The first try failed w/ mockito 3.4.0 because I ran into the [following](https://github.com/mockito/mockito/commit/12ba5936b736a2886220f9ae4a6492558dba4e14) issue.
Upgraded to the latest patch version of mockito (3.4.6) to have all such fixes.

Reply:
Seems like fixed all the issues and all the tests passed.